### PR TITLE
fix(java): handle formatted values in inner enum comparisons

### LINKED ISSUE #20952: Inner URI Enums with `useEnumCaseInsensitive` causes compilation errors
https://github.com/OpenAPITools/openapi-generator/blob/08ec37cc10dc1bb59ea541d27efbe8236c01883d/modules/openapi-generator/src/main/resources/Java/modelInnerEnum.mustache#L54

The current implementation of inner enum classes are not working for:
- `<library>okhttp-gson</library>`
- `<enumCaseInsensitive>true</enumCaseInsensitive>`
- `type: string`
- `format: uri`

This would generate:
```java
public static ExampleInnerThreeEnum fromValue(URI value) {
      for (ExampleInnerThreeEnum b : ExampleInnerThreeEnum.values()) {
        if (b.value.equalsIgnoreCase(value)) {
          return b;
        }
      }
      throw new IllegalArgumentException("Unexpected value '" + value + "'");
    }
```
Which will cause compilation errors, since `URI` does not have a method `equalsIgnoreCase`.

### PR DESCRIPTION
**Pull request by KannaKim**
### Description

Fixes #20952.

Formatted inner enums such as `type: string, format: uri` can generate `equalsIgnoreCase()` calls on non-String Java values, causing compilation errors when `useEnumCaseInsensitive=true`.

This change updates 11 affected inner-enum templates to compare formatted string enum values using toString() when useEnumCaseInsensitive=true. Unformatted strings retain their existing behavior, and comparisons with the option disabled remain unchanged.

Null inputs are handled without calling `toString()` on them. Existing equality behavior is preserved for non-string enums and when `useEnumCaseInsensitive=false`.

### Testing

All 26 parameterized cases pass across 13 generator/library configurations, with case-insensitive matching enabled and disabled.

Tests compile and execute generated `fromValue` methods in isolated enum fixtures, covering:

- Plain strings, email, and custom string formats.
- URI and UUID values.
- Unformatted and formatted numeric enums.
- Exact matches and differently cased inputs.
- Nullable and non-nullable enums, including distinguishing Java `null` from the literal string `"null"`.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package || exit
  ./bin/generate-samples.sh ./bin/configs/*.yaml || exit
  ./bin/utils/export_docs_generators.sh || exit
  ``` 
  (For Windows users, please run the script in [WSL](https://learn.microsoft.com/en-us/windows/wsl/install))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.
@bbdouglas (2017/07) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01) @karismann (2019/03) @Zomzog (2019/04) @lwlee2608 (2019/10) @martin-mfg (2023/08) @KannaKim (2026/07)

---
## Summary by cubic
Fixes #20952. Inner enum comparisons for string formats like `uri` no longer generate `equalsIgnoreCase()` calls on non-String values, so Java code using `useEnumCaseInsensitive=true` compiles.

- Updates 11 inner-enum templates to use `java.util.Objects.equals()` when a format is present.
- Unformatted string enums keep case-insensitive matching.
- Formatted string enums mapped to Java `String` now match case-sensitively.
- Adds parameterized regression tests that compile and run generated `fromValue` methods across 13 generator/library configurations.

<sup>Written for commit e0ecda9068dddcd1b8f68f4f7f21967121ea096e. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/OpenAPITools/openapi-generator/pull/24937?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/JavaJaxRS/cxf/enumClass.mustache [hunk 1]
```diff
@@ -32,7 +32,7 @@
     {{/jackson}}
     public static {{datatypeWithEnum}} fromValue({{dataType}} value) {
         for ({{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} b : {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}}.values()) {
-            if (b.value.equals{{#isString}}{{#useEnumCaseInsensitive}}IgnoreCase{{/useEnumCaseInsensitive}}{{/isString}}(value)) {
+            if ({{#isString}}{{#useEnumCaseInsensitive}}b.value{{#dataFormat}}.toString(){{/dataFormat}}.equalsIgnoreCase({{#dataFormat}}value == null ? null : value.toString(){{/dataFormat}}{{^dataFormat}}value{{/dataFormat}}){{/useEnumCaseInsensitive}}{{^useEnumCaseInsensitive}}b.value.equals(value){{/useEnumCaseInsensitive}}{{/isString}}{{^isString}}b.value.equals(value){{/isString}}) {
```
[the remark this answers is not in the document]
Comment (KannaKim) [reply modules/openapi-generator/src/main/resources/JavaJaxRS/cxf/enumClass.mustache:35]: `equalsIgnoreCase()` is gated by `useEnumCaseInsensitive=true`, so matching `FIRST` and `first` is intentional when that option is enabled. When disabled, the template uses `b.value.equals(value)`. The earlier PR description incorrectly implied that formatted Java strings should always use case-sensitive equality; this change preserves their existing case-insensitive behavior when enabled.

File: modules/openapi-generator/src/main/resources/JavaPlayFramework/enumClass.mustache [hunk 2]
```diff
@@ -35,7 +35,7 @@
     @JsonCreator
     public static {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} fromValue({{{dataType}}} value) {
       for ({{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} b : {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}}.values()) {
-        if (b.value.equals{{#isString}}{{#useEnumCaseInsensitive}}IgnoreCase{{/useEnumCaseInsensitive}}{{/isString}}(value)) {
+        if ({{#isString}}{{#useEnumCaseInsensitive}}b.value{{#dataFormat}}.toString(){{/dataFormat}}.equalsIgnoreCase({{#dataFormat}}value == null ? null : value.toString(){{/dataFormat}}{{^dataFormat}}value{{/dataFormat}}){{/useEnumCaseInsensitive}}{{^useEnumCaseInsensitive}}b.value.equals(value){{/useEnumCaseInsensitive}}{{/isString}}{{^isString}}b.value.equals(value){{/isString}}) {
```
[the remark this answers is not in the document]
Comment (KannaKim) [reply modules/openapi-generator/src/main/resources/JavaPlayFramework/enumClass.mustache:38]: similar reason as above, equalsIgnoreCase is gated by #useEnumCaseInsensitive. So case-insensitive comparison cannot happen unless useEnumCaseInsensitive=true.

File: modules/openapi-generator/src/main/resources/JavaSpring/enumClass.mustache [hunk 3]
```diff
@@ -53,7 +53,7 @@
     @JsonCreator
     public static {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} fromValue({{{dataType}}} value) {
       for ({{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} b : {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}}.values()) {
-        if (b.value.equals{{#isString}}{{#useEnumCaseInsensitive}}IgnoreCase{{/useEnumCaseInsensitive}}{{/isString}}(value)) {
+        if ({{#isString}}{{#useEnumCaseInsensitive}}b.value{{#dataFormat}}.toString(){{/dataFormat}}.equalsIgnoreCase({{#dataFormat}}value == null ? null : value.toString(){{/dataFormat}}{{^dataFormat}}value{{/dataFormat}}){{/useEnumCaseInsensitive}}{{^useEnumCaseInsensitive}}b.value.equals(value){{/useEnumCaseInsensitive}}{{/isString}}{{^isString}}b.value.equals(value){{/isString}}) {
```
[the remark this answers is not in the document]
Comment (KannaKim) [reply modules/openapi-generator/src/main/resources/JavaSpring/enumClass.mustache:56]: "When useEnumCaseInsensitive=true, formatted inner enums still use equalsIgnoreCase, so URI and other formatted values differing only by case resolve to the enum"

yes, it's intended behaviour.

File: modules/openapi-generator/src/main/resources/JavaJaxRS/enumClass.mustache [hunk 4]
```diff
@@ -49,7 +49,7 @@
     @JsonCreator
     public static {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} fromValue({{{dataType}}} value) {
       for ({{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} b : {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}}.values()) {
-        if (b.value.equals{{#isString}}{{#useEnumCaseInsensitive}}IgnoreCase{{/useEnumCaseInsensitive}}{{/isString}}(value)) {
+        if ({{#isString}}{{#useEnumCaseInsensitive}}b.value{{#dataFormat}}.toString(){{/dataFormat}}.equalsIgnoreCase({{#dataFormat}}value == null ? null : value.toString(){{/dataFormat}}{{^dataFormat}}value{{/dataFormat}}){{/useEnumCaseInsensitive}}{{^useEnumCaseInsensitive}}b.value.equals(value){{/useEnumCaseInsensitive}}{{/isString}}{{^isString}}b.value.equals(value){{/isString}}) {
```
[the remark this answers is not in the document]
Comment (KannaKim) [reply modules/openapi-generator/src/main/resources/JavaJaxRS/enumClass.mustache:52]: similar reason as above

File: modules/openapi-generator/src/main/resources/java-micronaut/common/model/modelInnerEnum.mustache [hunk 5]
```diff
@@ -46,7 +46,7 @@
     {{/jackson}}
         public static {{#datatypeWithEnum}}{{{.}}}{{/datatypeWithEnum}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} fromValue({{{dataType}}} value) {
             for ({{#datatypeWithEnum}}{{{.}}}{{/datatypeWithEnum}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} b : {{#datatypeWithEnum}}{{{.}}}{{/datatypeWithEnum}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}}.values()) {
-                if (b.value.equals{{#isString}}{{#useEnumCaseInsensitive}}IgnoreCase{{/useEnumCaseInsensitive}}{{/isString}}(value)) {
+                if ({{#isString}}{{#useEnumCaseInsensitive}}b.value{{#dataFormat}}.toString(){{/dataFormat}}.equalsIgnoreCase({{#dataFormat}}value == null ? null : value.toString(){{/dataFormat}}{{^dataFormat}}value{{/dataFormat}}){{/useEnumCaseInsensitive}}{{^useEnumCaseInsensitive}}b.value.equals(value){{/useEnumCaseInsensitive}}{{/isString}}{{^isString}}b.value.equals(value){{/isString}}) {
```
[the remark this answers is not in the document]
Comment (KannaKim) [reply modules/openapi-generator/src/main/resources/java-micronaut/common/model/modelInnerEnum.mustache:49]: intended behaviour

File: modules/openapi-generator/src/main/resources/Java/modelInnerEnum.mustache [hunk 6]
```diff
@@ -51,7 +51,7 @@
 {{/jackson}}
     public static {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} fromValue({{{dataType}}} value) {
       for ({{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} b : {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}}.values()) {
-        if (b.value.equals{{#isString}}{{#useEnumCaseInsensitive}}IgnoreCase{{/useEnumCaseInsensitive}}{{/isString}}(value)) {
+        if ({{#isString}}{{#useEnumCaseInsensitive}}b.value{{#dataFormat}}.toString(){{/dataFormat}}.equalsIgnoreCase({{#dataFormat}}value == null ? null : value.toString(){{/dataFormat}}{{^dataFormat}}value{{/dataFormat}}){{/useEnumCaseInsensitive}}{{^useEnumCaseInsensitive}}b.value.equals(value){{/useEnumCaseInsensitive}}{{/isString}}{{^isString}}b.value.equals(value){{/isString}}) {
```
[the remark this answers is not in the document]
Comment (KannaKim) [reply modules/openapi-generator/src/main/resources/Java/modelInnerEnum.mustache:54]: intended behaviour

File: modules/openapi-generator/src/main/resources/JavaJaxRS/spec/enumClass.mustache [hunk 7]
```diff
@@ -46,7 +46,7 @@
     @JsonCreator
     public static {{datatypeWithEnum}} fromValue({{dataType}} value) {
         for ({{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} b : {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}}.values()) {
-            if (b.value.equals{{#isString}}{{#useEnumCaseInsensitive}}IgnoreCase{{/useEnumCaseInsensitive}}{{/isString}}(value)) {
+            if ({{#isString}}{{#useEnumCaseInsensitive}}b.value{{#dataFormat}}.toString(){{/dataFormat}}.equalsIgnoreCase({{#dataFormat}}value == null ? null : value.toString(){{/dataFormat}}{{^dataFormat}}value{{/dataFormat}}){{/useEnumCaseInsensitive}}{{^useEnumCaseInsensitive}}b.value.equals(value){{/useEnumCaseInsensitive}}{{/isString}}{{^isString}}b.value.equals(value){{/isString}}) {
```
[the remark this answers is not in the document]
Comment (KannaKim) [reply modules/openapi-generator/src/main/resources/JavaJaxRS/spec/enumClass.mustache:49]: The reviewer is describing a pre-existing issue with enum constant initialization, which this PR doesn't modify. This fix specifically converts formatted enum values to Java strings via `toString()` inside `fromValue()` when `useEnumCaseInsensitive=true`, allowing `equalsIgnoreCase()` to be called on non-String-backed enums.

File: modules/openapi-generator/src/main/resources/JavaJaxRS/cxf-cdi/enumClass.mustache [hunk 8]
```diff
@@ -26,7 +26,7 @@
 
     public static {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} fromValue({{dataType}} value) {
         for ({{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} b : {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}}.values()) {
-            if (b.value.equals{{#isString}}{{#useEnumCaseInsensitive}}IgnoreCase{{/useEnumCaseInsensitive}}{{/isString}}(value)) {
+            if ({{#isString}}{{#useEnumCaseInsensitive}}b.value{{#dataFormat}}.toString(){{/dataFormat}}.equalsIgnoreCase({{#dataFormat}}value == null ? null : value.toString(){{/dataFormat}}{{^dataFormat}}value{{/dataFormat}}){{/useEnumCaseInsensitive}}{{^useEnumCaseInsensitive}}b.value.equals(value){{/useEnumCaseInsensitive}}{{/isString}}{{^isString}}b.value.equals(value){{/isString}}) {
```
[the remark this answers is not in the document]
Comment (KannaKim) [reply modules/openapi-generator/src/main/resources/JavaJaxRS/cxf-cdi/enumClass.mustache:29]: resolving conversation with similar reasoning as the reviews above

File: modules/openapi-generator/src/main/resources/JavaJaxRS/spec/enumClass.mustache (hunk 7 above)
[the remark this answers is not in the document]
Comment (KannaKim) [reply modules/openapi-generator/src/main/resources/JavaJaxRS/spec/enumClass.mustache:49]: resolving conversation with similar reasoning as the reviews above

File: modules/openapi-generator/src/main/resources/Java/libraries/okhttp-gson/modelInnerEnum.mustache [hunk 9]
```diff
@@ -39,7 +39,7 @@
 
     public static {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} fromValue({{{dataType}}} value) {
       for ({{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}} b : {{{datatypeWithEnum}}}{{^datatypeWithEnum}}{{{classname}}}{{/datatypeWithEnum}}.values()) {
-        if (b.value.equals{{#isString}}{{#useEnumCaseInsensitive}}IgnoreCase{{/useEnumCaseInsensitive}}{{/isString}}(value)) {
+        if ({{#isString}}{{#useEnumCaseInsensitive}}b.value{{#dataFormat}}.toString(){{/dataFormat}}.equalsIgnoreCase({{#dataFormat}}value == null ? null : value.toString(){{/dataFormat}}{{^dataFormat}}value{{/dataFormat}}){{/useEnumCaseInsensitive}}{{^useEnumCaseInsensitive}}b.value.equals(value){{/useEnumCaseInsensitive}}{{/isString}}{{^isString}}b.value.equals(value){{/isString}}) {
```
[the remark this answers is not in the document]
Comment (KannaKim) [reply modules/openapi-generator/src/main/resources/Java/libraries/okhttp-gson/modelInnerEnum.mustache:42]: resolving conversation with similar reasoning as the reviews above

**Comment by KannaKim** [comment]
The cubic review was probably based on my previous PR description. I’ve updated the description to accurately reflect the implementation.

File: modules/openapi-generator/src/test/resources/3_0/issue_20952_inner_enum_comparison.yaml [hunk 10]
```diff
@@ -0,0 +1,53 @@
+openapi: 3.0.3
```
Comment (wing328) [inline modules/openapi-generator/src/test/resources/3_0/issue_20952_inner_enum_comparison.yaml:1]: FYI. Tested with this (using the default `okhttp-gson` library) but the output won't compile:

```
[ERROR] C:\Users\User\AppData\Local\Temp\enumtest2\src\main\java\org\openapitools\client\model\Example.java:[321,18] error: no suitable method found for value(UUID)
[ERROR]     method JsonWriter.value(String) is not applicable
[ERROR]       (argument mismatch; UUID cannot be converted to String)
[ERROR]     method JsonWriter.value(boolean) is not applicable
[ERROR]       (argument mismatch; UUID cannot be converted to boolean)
[ERROR]     method JsonWriter.value(Boolean) is not applicable
[ERROR]       (argument mismatch; UUID cannot be converted to Boolean)
[ERROR]     method JsonWriter.value(float) is not applicable
[ERROR]       (argument mismatch; UUID cannot be converted to float)
[ERROR]     method JsonWriter.value(double) is not applicable
[ERROR]       (argument mismatch; UUID cannot be converted to double)
[ERROR]     method JsonWriter.value(long) is not applicable
[ERROR]       (argument mismatch; UUID cannot be converted to long)
[ERROR]     method JsonWriter.value(Number) is not applicable
[ERROR]       (argument mismatch; UUID cannot be converted to Number)
[ERROR] C:\Users\User\AppData\Local\Temp\enumtest2\src\main\java\org\openapitools\client\model\Example.java:[326,32] error: cannot find symbol
[ERROR]   symbol:   method nextUUID()
[ERROR]   location: variable jsonReader of type JsonReader
[ERROR] C:\Users\User\AppData\Local\Temp\enumtest2\src\main\java\org\openapitools\client\model\Example.java:[332,30] error: cannot find symbol
[ERROR]   symbol:   method getAsUUID()
[ERROR]   location: variable jsonElement of type JsonElement

```` 

Not related to this change. Just another edge case we need to address one day.

**Comment by wing328** [comment]
thanks for the PR.

I did a test with

```
java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -g java -i https://raw.githubusercontent.com/KannaKim/openapi-generator/1a102a1226560e5c7482f2bb1714e2fff877305e/modules/openapi-generator/src/test/resources/3_0/issue_20952_inner_enum_comparison.yaml -o /tmp/enumtest2/
```

But don't see any change when compared with the latest `master`.

Are you seeing the same or I'm no testing this change correctly ?

File: modules/openapi-generator/src/test/resources/3_0/issue_20952_inner_enum_comparison.yaml (hunk 10 above)
Comment (KannaKim) [reply modules/openapi-generator/src/test/resources/3_0/issue_20952_inner_enum_comparison.yaml:1]: Noted, I can make separate PR for this.

**Comment by KannaKim** [comment]
> thanks for the PR.
> 
> I did a test with
> 
> ```
> java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -g java -i https://raw.githubusercontent.com/KannaKim/openapi-generator/1a102a1226560e5c7482f2bb1714e2fff877305e/modules/openapi-generator/src/test/resources/3_0/issue_20952_inner_enum_comparison.yaml -o /tmp/enumtest2/
> ```
> 
> But don't see any change when compared with the latest `master`.
> 
> Are you seeing the same or I'm no testing this change correctly ?

Hi, thanks for testing. It looks like _useEnumCaseInsensitive=true_ was omitted in your command. I just ran test on my end using 
```
java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -g java -i https://raw.githubusercontent.com/KannaKim/openapi-generator/1a102a1226560e5c7482f2bb1714e2fff877305e/modules/openapi-generator/src/test/resources/3_0/issue_20952_inner_enum_comparison.yaml -o /tmp/enumtest2/ --additional-properties=useEnumCaseInsensitive=true
```
generated comparison changed from 
```java
b.value.equalsIgnoreCase(value)
```
to
```java
b.value.toString().equalsIgnoreCase(value == null ? null : value.toString())
```

**Comment by wing328** [comment]
Ah ok. I missed that. Thanks for the details and the fix

File: modules/openapi-generator/src/main/resources/JavaSpring/enumClass.mustache (hunk 3 above)
[the remark this answers is not in the document]
Comment (Chrimle) [reply modules/openapi-generator/src/main/resources/JavaSpring/enumClass.mustache]: Don't see the point in these null checks and so on. The fix should be a one-liner:
https://github.com/Chrimle/openapi-to-java-records-mustache-templates/commit/605597c0184cffeb3f2bdabf50205a3156665e7a#diff-aeb7136d23f600e2b94b74af6d963df0114170da689c37a60bc64bbccea4565cL140

Also, no sample files to see how/what this generates?
